feat(new_sink): initial `websocket_listener` sink

esensar opened the pull request:
## Summary

This adds a new sink, similar to `websocket` which acts like a client, while this one acts as a server. It currently broadcasts messages to all currently connected clients. 

It might make sense to add some kind of buffer and allow clients to catch up to missed messages, or something like that, but I wanted to leave such complications for further PRs. 

This also adds customizable authentication using VRL, which would allow us to use custom VRL scripts to check auth, meaning we no longer use hardcoded basic auth and can use enrichment tables to look for credentials.

## Change Type
- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## How did you test this PR?

Started up vector with the following configuration:
```yaml
sources:
  demo_logs_test:
    type: "demo_logs"
    format: "json"

sinks:
  websocket_sink:
    inputs: ["demo_logs_test"]
    type: "websocket_listener"
    address: "0.0.0.0:1234"
    encoding:
      codec: "json"
```

Connected to the server with [websocat](https://github.com/vi/websocat) and observed that events are sent to the client. I have also tested auth in this way, by defining auth header using `websocat`.

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist
- [x] Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
  - `make check-all` is a good command to run locally. This check is
    defined [here](https://github.com/vectordotdev/vector/blob/1ef01aeeef592c21d32ba4d663e199f0608f615b/Makefile#L450-L454). Some of these
    checks might not be relevant to your PR. For Rust changes, at the very least you should run:
    - `cargo fmt --all`
    - `cargo clippy --workspace --all-targets -- -D warnings`
    - `cargo nextest run --workspace` (alternatively, you can run `cargo test --all`)
- [ ] If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `dd-rust-license-tool write` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

esensar commented:
I have added customizable auth handler in this PR, but I can take it out into a separate PR if needed.

This is also missing some tests and some metrics. I will probably also work on an optional event buffer to enable replays for some clients. That one might be more complex, so I wanted to put it off for now, to keep the initial PR simple.

pront commented:
> I have added customizable auth handler in this PR, but I can take it out into a separate PR if needed.
Are you referring to src/common/server_auth.rs? I also wonder if that code adds duplication with other http components.

If it's not too much effort, please do.  It should make reviewing faster. It's just me reviewing PRs for this month 😅

esensar commented:
> > I have added customizable auth handler in this PR, but I can take it out into a separate PR if needed.
> > Are you referring to src/common/server_auth.rs? I also wonder if that code adds duplication with other http components.
> 
> If it's not too much effort, please do. It should make reviewing faster. It's just me reviewing PRs for this month 😅

Oh :smile: 
Alright, I will do it, it should be pretty simple to move out.

esensar commented:
Oh and sorry, I just saw the other part of the comment. Yeah, it causes duplication, my idea was to reuse it across these components, but didn't want to make too many changes in this PR. I guess if I move it to a separate PR, then changing these out should be okay.

esensar commented:
Would it make more sense to name this `websocket_server`? Not sure why I named it listener to be honest. I think `server` would be more consistent with other components.

jszwedko commented:
> Would it make more sense to name this `websocket_server`? Not sure why I named it listener to be honest. I think `server` would be more consistent with other components.

👍 `websocket_server` would be consistent with `http_server`. It leaves the door open for a `websocket_client` source that, like the `http_client` source, would initiate requests rather than listen for them.

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+			description: """
+				Whether or not end-to-end acknowledgements are enabled.
+
+				When enabled for a sink, any source connected to that sink, where the source supports
+				end-to-end acknowledgements as well, waits for events to be acknowledged by **all
+				connected** sinks before acknowledging them at the source.
```
```suggestion
				When enabled for a sink, any source that supports end-to-end
				acknowledgements that is connected to that sink waits for events
				to be acknowledged by **all connected sinks** before acknowledging
				them at the source.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+		description: """
+			The socket address to listen for connections on.
+
+			It _must_ include a port.
```
```suggestion
			This value _must_ include a port.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+					}
+					device_version: {
+						description: """
+																Identifies the version of the problem. In combination with device product and vendor, it composes the unique id of the device that sends messages.
```
```suggestion
																Identifies the version of the problem. The combination of the device produc, vendor,  and this value make up the unique id of the device that sends messages.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+					name: {
+						description: """
+																This is a path that points to the human-readable description of a log event.
+																The value length must be less than or equal to 512.
```
Juts checking - 512 and not 511?

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+																Reflects importance of the event.
+
+																It must point to a number from 0 to 10.
+																0 = Lowest, 10 = Highest.
```
```suggestion
																0 = lowest importance, 10 = highest importance.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+
+																It must point to a number from 0 to 10.
+																0 = Lowest, 10 = Highest.
+																Equals to "cef.severity" by default.
```
```suggestion
																Set to `cef.severity` by default.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+					version: {
+						description: """
+																CEF Version. Can be either 0 or 1.
+																Equals to "0" by default.
```
```suggestion
																Set to `0` by default.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+						Vector's encoder currently adheres more strictly to the GELF spec, with
+						the exception that some characters such as `@`  are allowed in field names.
+
+						Other GELF codecs such as Loki's, use a [Go SDK][implementation] that is maintained
```
```suggestion
						Other GELF codecs, such as Loki's, use a [Go SDK][implementation] that is maintained
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+						the exception that some characters such as `@`  are allowed in field names.
+
+						Other GELF codecs such as Loki's, use a [Go SDK][implementation] that is maintained
+						by Graylog, and is much more relaxed than the GELF spec.
```
```suggestion
						by Graylog and is much more relaxed than the GELF spec.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+						Other GELF codecs such as Loki's, use a [Go SDK][implementation] that is maintained
+						by Graylog, and is much more relaxed than the GELF spec.
+
+						Going forward, Vector will use that [Go SDK][implementation] as the reference implementation, which means
```
Can you define "going forward" more specifically?

esensar replied:
Since this is part of general documentation which is just generated here, I am not sure what is exactly meant here, so it would probably be better for somebody else to define this more specifically.

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+						by Graylog, and is much more relaxed than the GELF spec.
+
+						Going forward, Vector will use that [Go SDK][implementation] as the reference implementation, which means
+						the codec may continue to relax the enforcement of specification.
```
```suggestion
						the codec might continue to relax the enforcement of the specification.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+				type: object: options: {
+					capacity: {
+						description: """
+																Set the capacity (in bytes) of the internal buffer used in the CSV writer.
```
```suggestion
																Sets the capacity (in bytes) of the internal buffer used in the CSV writer.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+					capacity: {
+						description: """
+																Set the capacity (in bytes) of the internal buffer used in the CSV writer.
+																This defaults to a reasonable setting.
```
I highly recommend specifying this as a value instead of using the words "reasonable setting", which are barely more specific than saying nothing.

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+					}
+					double_quote: {
+						description: """
+																Enable double quote escapes.
```
```suggestion
																Enables double quote escapes.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+						description: """
+																Enable double quote escapes.
+
+																This is enabled by default, but it may be disabled. When disabled, quotes in
```
```suggestion
																This is enabled by default, but you can disable it. When disabled, quotes in
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+																In some variants of CSV, quotes are escaped using a special escape character
+																like \\ (instead of escaping quotes by doubling them).
+
+																To use this, `double_quotes` needs to be disabled as well otherwise it is ignored.
```
```suggestion
																To use this, `double_quotes` needs to be disabled as well; otherwise, this setting is ignored.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+					}
+					fields: {
+						description: """
+																Configures the fields that will be encoded, as well as the order in which they
```
```suggestion
																Configures the fields that are encoded, as well as the order in which they
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+																Configures the fields that will be encoded, as well as the order in which they
+																appear in the output.
+
+																If a field is not present in the event, the output will be an empty string.
```
```suggestion
																If a field is not present in the event, the output for that field is an empty string.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+
+																If a field is not present in the event, the output will be an empty string.
+
+																Values of type `Array`, `Object`, and `Regex` are not supported and the
+																output will be an empty string.
```
```suggestion
																Values of type `Array`, `Object`, and `Regex` are not supported, and the
																output for any of these types is an empty string.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+								never: "Never writes quotes, even if it produces invalid CSV data."
+								non_numeric: """
+																			Puts quotes around all fields that are non-numeric.
+																			Namely, when writing a field that does not parse as a valid float or integer,
```
```suggestion
																			This means that when writing a field that does not parse as a valid float or integer,
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+								non_numeric: """
+																			Puts quotes around all fields that are non-numeric.
+																			Namely, when writing a field that does not parse as a valid float or integer,
+																			then quotes are used even if they aren't strictly necessary.
```
```suggestion
																		quotes are used even if they aren't strictly necessary.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+				description: """
+					Controls how metric tag values are encoded.
+
+					When set to `single`, only the last non-bare value of tags are displayed with the
```
```suggestion
					When set to `single`, only the last non-bare value of tags is displayed with the
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+				description: """
+					Sets the list of supported ALPN protocols.
+
+					Declare the supported ALPN protocols, which are used during negotiation with peer. They are prioritized in the order
```
```suggestion
					Declare the supported ALPN protocols, which are used during negotiation with a peer. They are prioritized in the order
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+					The certificate must be in DER, PEM (X.509), or PKCS#12 format. Additionally, the certificate can be provided as
+					an inline string in PEM format.
+
+					If this is set, and is not a PKCS#12 archive, `key_file` must also be set.
```
```suggestion
					If this is set _and_ is not a PKCS#12 archive, `key_file` must also be set.
```

neko-dd commented on website/cue/reference/components/sinks/base/websocket_server.cue:
```diff
@@ -0,0 +1,501 @@
+					If enabled, certificates must not be expired and must be issued by a trusted
+					issuer. This verification operates in a hierarchical manner, checking that the leaf certificate (the
+					certificate presented by the client/server) is not only valid, but that the issuer of that certificate is also valid, and
+					so on until the verification process reaches a root certificate.
```
```suggestion
					so on, until the verification process reaches a root certificate.
```

esensar commented:
@neko-dd sorry for the late reply. I have applied on of the suggestions, which was directly related to things added in this PR. The rest of it has been generated automatically from other things that I have used (which are standard for all sinks) and you can see that documentation already live on almost all of the sinks. Maybe these should be fixed in a separate PR? Because if I fixed the here, I would generate too many changes that are not really related to this (it would change docs for all sinks).

pront commented:
> Maybe these should be fixed in a separate PR? Because if I fixed the here, I would generate too many changes that are not really related to this (it would change docs for all sinks).

Good idea 👍 

---

Also, please take a look at this 'guide', we will need some more doc files:
https://github.com/vectordotdev/vector/pull/22072#issuecomment-2605051340

esensar commented:
> Also, please take a look at this 'guide', we will need some more doc files: [#22072 (comment)](https://github.com/vectordotdev/vector/pull/22072#issuecomment-2605051340)

Does this look better now? I have based this on `websocket` docs.

pront reviewed (approved):
I did another pass and this looks good!

esensar commented:
@pront is this okay to merge? I have updated it with changes from #22236

pront commented:
> @pront is this okay to merge? I have updated it with changes from #22236

Hi @esensar, there are open comments around docs and also I need to do another review before merging.

esensar commented:
> > @pront is this okay to merge? I have updated it with changes from #22236
> 
> Hi @esensar, there are open comments around docs and also I need to do another review before merging.

I have added the docs changes in a separate PR: https://github.com/vectordotdev/vector/pull/22317

pront commented:
> I have added the docs changes in a separate PR: #22317

Ah yes, this is coming back to me 👍. Did we merge the master with those changes in this PR? If yes, please resolve those comments.

pront commented on src/internal_events/websocket_server.rs:
```diff
@@ -0,0 +1,71 @@
+
+impl InternalEvent for WsListenerConnectionEstablished {
+    fn emit(self) {
+        debug!(message = "Websocket client connected.");
```
Nit 
```suggestion
        debug!(message = "Websocket client connected. Client count: {self.client_count}");
```

pront commented on src/internal_events/websocket_server.rs:
```diff
@@ -0,0 +1,71 @@
+
+impl InternalEvent for WsListenerConnectionShutdown {
+    fn emit(self) {
+        warn!(message = "Client connection closed.");
```
Is this a warning or an info?

Nit:
```suggestion
        warn!(message = "Client connection closed. Client count: {self.client_count}.");
```

esensar replied:
Maybe info would make more sense. I took websocket events as an example here and it used warn, but maybe it makes more sense for a client. I guess a lot more disconnections are expected on a server.

esensar commented:
> > I have added the docs changes in a separate PR: #22317
> 
> Ah yes, this is coming back to me 👍. Did we merge the master with those changes in this PR? If yes, please resolve those comments.

Yes, but now looking at them, it seems that I have missed some of them. I will open up another PR with these.

esensar commented:
> > > I have added the docs changes in a separate PR: #22317
> > 
> > 
> > Ah yes, this is coming back to me 👍. Did we merge the master with those changes in this PR? If yes, please resolve those comments.
> 
> Yes, but now looking at them, it seems that I have missed some of them. I will open up another PR with these.

Actually, it seems that I just forgot to regenerate documentation. It should all be alright now.

esensar commented:
> Wonderful, happy to see this merged. Optional request @esensar: can we add a "How it works" section with the steps you outlined in the test plan? The community has been finding these useful.

Of course! I will add that in tomorrow. Do you think something like that would be useful for https://github.com/vectordotdev/vector/pull/22236 too? Or is the included documentation enough.

pront commented:
> > Wonderful, happy to see this merged. Optional request @esensar: can we add a "How it works" section with the steps you outlined in the test plan? The community has been finding these useful.
> 
> Of course! I will add that in tomorrow. Do you think something like that would be useful for #22236 too? Or is the included documentation enough.

Thanks! Yes, more usage examples are important. They help reduce user frustration and also our KTLO cost.
